### DISCUSSION TITLE
IO data updates

### DISCUSSION THREAD

**Discussion by WesIngwersen** [USEPA/useeior#275] [General]
Discuss strategies and issues related to updating useeior for updated input data

**Comment by WesIngwersen** [comment]
The BEA benchmark IO and underlying data release which are every 5 yrs is the most significant regular data update required. This update comes with a schema change with new commodities and industries.

Functions to retrieve the data are in data-raw. In BEAData. R they are currently tied to a single year in name. Also they are very repetitive.. more like a copy and paste of each other.

**Reply by bl-young** [reply]
I would propose using an updated naming schema for data files from BEA whereby each filename is appended with a tag for the BEA schema and a subtag that includes a 2-digit numeric version that aligns with the specific BEA release: `{year}.{v}`

i.e.:
`Detail_GrossOutput_IO_2012.02.rda` or `Detail_Use_2012_PRO_AfterRedef_2012.02.rda` or `Summary_Make_2019_AfterRedef_2017.01.rda`

We could log within the package the alignment between data versions and BEA releases by date for any given schema year.

In this way we might have versions of the Summary_Make table for 2017 under both the 2012 schema and 2017 schema, the version number that is the highest would reflect the most recent release by BEA.

**Reply by bl-young** [reply]
First pass at consolidating the functions (detail tables to start) is in https://github.com/USEPA/useeior/commit/cc91a022bb545c3d9d6c1c73f4af7900bb2b8bda. They can be consolidated further but doing it piecemeal.

**Reply by WesIngwersen** [reply]
Might want a letter(s) appended to the year representing the schema in that name like 2017sch to distinguish it from the data year but otherwise a good proposal

**Reply by bl-young** [reply]
yes ok, in fact I was also thinking we could abbreviate the schema to two digits, so:
`Summary_Make_2019_AfterRedef_17sch.01.rda`

**Reply by bl-young** [reply]
@WesIngwersen one hiccup with this schema - I went ahead and [renamed ](https://github.com/USEPA/useeior/commit/131f974e7e3d6ba2912521cad9e3c4314f61b01e)the 2012 schema file as planned, e.g., `Summary_Make_2014_AfterRedef_12sch.rda`. However, these .rda files are called by the name of the object when they are saved, not by the name of the file apparently. So after renaming, you still have to use: `get('Summary_Make_2014_AfterRedef.rda')`. I think there are a few options:

1. keep the name change of the 2012 files, so in the repo you can see which data file is which based on the file name, knowing that this will cause a discrepancy between the filename and what is called with `get()`
2. revert the name change for the 2012 files so they are the same as before, when we adjust the code for getting IO data, append the schema only when the schema is not 2012.
3. I could probably write a script to load all the 2012 schema files, rename the object, and write back to RDA with the new schema name.

**Reply by bl-young** [reply]
discussed ok to stick with option 1

**Comment by bl-young** [comment]
Regarding the data objects and metadata files that get created. Currently they get stored in `data/` and `inst/ext/metadata` respectively. Storing the data files uses `usethis::use_data`

https://github.com/USEPA/useeior/blob/c8849037b166edb75f4588257d899826994bc6cf/R/UtilityFunctions.R#L378-L383

Would there be any benefit to storing these data objects on data commons / appdirs. Given their small size I'm not seeing major benefit to making that shift, especially once the new schema are in place.

In fact I'm even reconsidering whether we need to indicate the "version number" in the file naming system (which would result in new files instead of replacing data files upon data updates). The benefit to version number would be if we want to be able to recreate prior versions of the model moving forward without having to back in the package history, otherwise we have all that information in the file metadata.

**Reply by bl-young** [reply]
Discussed that keeping in the repo and replacing files makes sense and is the path forward. Will want to be explicit when data are updated to note those updates in package releases.

**Comment by bl-young** [comment]
Another question: do we want to enable access to be able to freshly obtain 2012 files from the BEA archive? Or can we assume that we won't need to do that anymore and just keep our processed files in the repository?

The former would require making further updates to enable pointing to the archive and adjusting some steps for 2012 schema tables.

**Reply by bl-young** [reply]
Confirmed - no explicit need to access the 2012 archive, though models for 2012 schema can still be built.

**Comment by bl-young** [comment]
There is now a [2017 schema sample model](https://github.com/USEPA/useeior/commit/6f157aa93a5bd6e1e68f0efef33ad78c65ef9ab3) (w/ GHGs) on that branch. You can see in the testing that it successfully creates the model, though [there are some validation issues](https://github.com/USEPA/useeior/actions/runs/7951153938/job/21704242299).

There are a few spots where I took some quick measures to get it working that I will need to go back to review (specifically around some mapping files when pulling the CPI and GO data, e.g. [here](https://github.com/USEPA/useeior/blob/6f157aa93a5bd6e1e68f0efef33ad78c65ef9ab3/data-raw/BEAData.R#L613-L620)). I expect those may be causing the validation errors but i will be looking closer.

**Reply by bl-young** [reply]
last remaining [validation issue](https://github.com/USEPA/useeior/actions/runs/7967224309/job/21749681015#step:8:54) is with `333914/US` Pumping Equipment Manufacturing. May be related to a mapping issue. is this a new BEA sector??

**Reply by WesIngwersen** [reply]
Confirmed yes

**Reply by bl-young** [reply]
the validation issue is now resolved, was due to mis-aligned sectors because commodity names were [missing for two new BEA sectors](https://github.com/USEPA/useeior/commit/017a5c18297cc396cc8ed5972e570ce83c701e47)

**Comment by bl-young** [comment]
**Regarding differences in NAICS schema for satellite tables compared to IO schema**, for example, the 2017 Model being tested using the published GHG datasets, which were built with 2012 NAICS schema

see the [new warning](https://github.com/USEPA/useeior/actions/runs/8084234383/job/22089066903#step:8:26) in model build: `WARNING::Lost sectors due to missing mapping: 333911, 333913, 335221, 335222, 335224, 335228, 5171, 5172`
In this case, 333911 and 333913 have been updated to 333914 in the 2017 NAICS (and identical BEA code). Because 333911 and 333913 no longer exist in the 2017 schema, those data points get dropped from the TbS.

Do we want to allow satellite tables w/ 2012 NAICS to be used in 2017 BEA models? If so we will need a function to do that conversion, similar to [`mapFlowTotalsbySectorfromBEASchema2007to2012`](https://github.com/USEPA/useeior/blob/8e8de0169c6f62edf7f42cbb86514a9ee08d2795/R/SatelliteFunctions.R#L248)

**Reply by WesIngwersen** [reply]
It would be cleaner not to allow this ... to require the schemas to be the same or else through an error.  I understand this would require rebuilding fbs models in some cases.. let's evaluate the trade-offs

**Reply by bl-young** [reply]
I think that's a fine approach for now. I'll note that because the new 2017 schema is currently only available from 2017-2022, many of our FBS wouldn't work anyway (e.g., water, land, etc.) that are from 2016 or earlier (there is no industry output for those years).
